### PR TITLE
[service/extensions] extension lifecycle order

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
Enforce order of start and shutdown of extensions according to configuration

**Link to tracking Issue:**
Fixes issue #8732

**Testing:**
Unit tests

### CODE REVIEW

**Review by yurishkuro — APPROVED** [review]
Thanks!

**Review by dmitryax — CHANGES_REQUESTED** [review]
Just to hold off until discussion in https://github.com/open-telemetry/opentelemetry-collector/issues/8732 is resolved

**Comment by atoulme** [comment]
I received a comment that I should stop extensions in reverse order that they are started, I'll push that. After that, we can discuss the merits of this PR.

**Comment by yurishkuro** [comment]
Another implementation option discussed on the issue is to build support for extension dependencies in the graph/ module. Then an extension can programmatically declare a dependency on another extension and the graph would guarantee correct execution order. This would be a larger change, but preferable over explicit config order where users need to worry about listing extensions in the right order.

**Comment by dmitryax** [comment]
> Another implementation option discussed on the issue is to build support for extension dependencies in the graph/ module. Then an extension can programmatically declare a dependency on another extension and the graph would guarantee correct execution order. This would be a larger change, but preferable over explicit config order where users need to worry about listing extensions in the right order.

Makes sense, but it would require another predefined way to specify dependencies on the extensions in the configuration interface. The current dependencies topology is constructed directly from the pipeline YAML definition. Introducing an additional source for building the graph seems like an excessive complication of the user interface. And some extensions naturally do not provide any capabilities to other components, only to the end-user. I think it's ok to keep it simple like implemented in this PR going forward.

**Comment by yurishkuro** [comment]
> Introducing an additional source for building the graph seems like an excessive complication of the user interface.

Don't get me wrong, I am happy with this PR. But when we talk about complicating user interface, I think it's important to remember _which users are exposed to such interface_. With dependencies expressed programmatically, a very small number of users who write their own extensions are affected. With dependencies expressed via config order, _all_ users of the collector are exposed to the notion that there are dependencies between extensions, even though there's very little they can actually do about it. E.g. `jaeger_query` extension simply will never work if it's placed ahead of `jaeger_storage` in the configuration. So in my documentation I will have to mention "always put it last", even though users didn't really need to be exposed to this accidental complexity.

**Comment by dmitryax** [comment]
@yurishkuro, how do you envision a programmatic way to define the dependencies? Something like an optional interface for extensions to define a dependency on another extension interface? I believe that doesn't require integration with the pipeline graph. We would still start all the extensions before the pipelines.

**Comment by yurishkuro** [comment]
> how do you envision a programmatic way to define the dependencies?

I'd add an optional argument `dependsOn []component.ID` here:

https://github.com/open-telemetry/opentelemetry-collector/blob/6405e152a2d9a08de1fe1b70fd4ac775a12a3cda/extension/extension.go#L103-L107

I'm not sure why these constructor APIs are currently not made extensible by taking args as a struct instead of fixed positional args.

**Comment by dmitryax** [comment]
> I'd add an optional argument dependsOn []component.ID here

In that case, you can only define a hardcoded dependency on a particular component. There should be a way to define a dependency on a class of extensions like storage or encoding, which is defined by an interface an extension implements.

**Comment by yurishkuro** [comment]
> In that case, you can only define a hardcoded dependency on a particular component. There should be a way to define a dependency on a class of extensions like storage or encoding, which is defined by an interface an extension implements.

I personally would avoid duck-typing dependencies, it's better to have explicit ones by ID. There is only one pool of extensions, so what if I have some components depending on one kind of encoding extension and other components depending on another? When they refer to those by ID there is no ambiguity, but duck-typing only works when there is exactly one instance defined for a given interface.

**Comment by dmitryax** [comment]
Extensions should provide the functionality to components via interfaces like [storage](https://github.com/open-telemetry/opentelemetry-collector/blob/main/extension/experimental/storage/storage.go#L14). Components dependent on that interface can choose any storage extension even if it's implemented outside of OTel core/contrib repos. Choosing a particular storage implementation is exposed to the user with a config option. Having a hardcoded set of IDs means that custom extensions of particular classes cannot be implemented outside, and every extension dependent on a particular class of extension has to maintain a set of all instances, e.g., each time a new storage/encoding extension is added, all dependent components have to be updated.

> There is only one pool of extensions, so what if I have some components depending on one kind of encoding extension and other components depending on another? 

They should be two classes of extensions that implement two different (or additional) interfaces.

**Comment by yurishkuro** [comment]
@dmitryax I didn't mean that the dependency **has** to be hardcoded: for some extensions it may be sufficient, others may need to express their dependencies after reading their own config. That means the dependencies should not be exposed via `NewFactory`, but as part of `CreateExtension` constructor. You're right, in this case a runtime interface check is a good solution - here's a prototype #8768 (to be combined with this PR for the actual ordered start/shutdown)

File: service/extensions/extensions.go [hunk 1]
```diff
@@ -77,7 +82,8 @@ func (bes *Extensions) NotifyPipelineReady() error {
 func (bes *Extensions) NotifyPipelineNotReady() error {
 	// Notify extensions in reverse order.
```
Comment (yurishkuro) [inline service/extensions/extensions.go:83]: is this accurate?

**Comment by dmitryax** [comment]
@yurishkuro, I think we can merge this one and rebase https://github.com/open-telemetry/opentelemetry-collector/pull/8768 on top of it since it won't cause any breaking changes, right?

File: service/extensions/extensions.go (hunk 1 above)
Comment (dmitryax) [reply service/extensions/extensions.go:83]: Looks like it's not. @atoulme can you please address this?

**Comment by yurishkuro** [comment]
> I think we can merge this one and rebase https://github.com/open-telemetry/opentelemetry-collector/pull/8768 on top of it since it won't cause any breaking changes, right?

@dmitryax my PR is already based on this one (it includes the original commits). I don't have a strong objection to merging each one separately, but it means the 1st PR will introduce one way of ordering (based purely on config order) and 2nd will override that in favor of the order defined by the DependentExtension interface (with other order still being random). In theory it's possible to support both at the same time: by default start with config-based order as documentation implies, but if DependentExtension interfaces are detected in some extensions move them to the end and apply topological sort (seems a bit complicated). I think just the 2nd approach is cleaner - no order guarantees unless explicitly requested via interface.

File: service/extensions/extensions.go (hunk 1 above)
Comment (atoulme) [reply service/extensions/extensions.go:83]: How would you like me to address this? It feels like this comment needs to be removed.

**Comment by dmitryax** [comment]
> I think just the 2nd approach is cleaner - no order guarantees unless explicitly requested via interface.

Ok, makes sense to me. Especially if that is simpler

**Comment by dmitryax** [comment]
Superseded by https://github.com/open-telemetry/opentelemetry-collector/pull/8768

**Pull request closed without merging**